src/client.py

```diff
@@ -0,0 +1,362 @@
+        if token:
+            self.authenticated = True
+        # Initializes requests Session and assigns headers for all Client MERMAID API calls.
+        self.session = requests.Session()
+        self.session.headers.update({
+            'content-type': "application/json",
+            'authorization': "Bearer %s" % self.token
+
+        })
+
+    # API paths
+    def api_root(self, subdirectory='', method='GET', parameters=None, data=None):
+        """
+        Prepares API call and utilizes Client Session to make MERMAID API calls.
+        :param subdirectory: (optional) subdirectory path. Defaults to (subdirectory='').
+        :type subdirectory: str
+        :param method: HTTP request method. Defaults to (method='GET').
+        :type method: str
+        :param parameters: (optional) parameters for request.
+        :type parameters: dict: (../?key=val), str: (../?str).
+        :param data: (optional) data. Defaults to (data=None).
+        :return: JSON object containing MERMAID API data.
+        :rtype: dict, None
+        """
+        # Creates full URL path.
+        prep_url = self.url + subdirectory
+        # Prepares Request and send from Client class Session.
+        req = requests.Request(method, url=prep_url, params=parameters, data=data)
+        prepped = self.session.prepare_request(req)
+        resp = self.session.send(prepped)
+
+        # Returns JSON if response code OK.
+        if resp.status_code == requests.codes.ok:
+            return resp.json()
+        else:
+            if resp.status_code == 401:
+                print('api_root response code: ' + str(resp.status_code) + ' -- Attempt token refresh')
+            else:
+                print('api_root response code: ' + str(resp.status_code))
+            return None
+
+    def api_projects(self, name=None, id=None, resource=None, filter=None, filter_val=None):
+        """
+        Helper class function used to return projects path '/projects/<project_id>/' based on name or id of project.
+        See https://mermaid-api.readthedocs.io/en/latest/projects.html#project-resources for more information.
+        :param name: (optional if 'id' provided) MERMAID project name.
+        :type name: str
+        :param id: (optional if 'name' provided) MERMAID project ID.
+        :type: id: str
+        :param filter: MERMAID project resource filters.
+        :type filter: str
+        :param filter_val: (optional) Required for resource filters requiring values. eg. (size_min/size_max,
+        count_min/count_max, length_min/length_max, len_surveyed_min/len_surveyed_max,
+        sample_date_before/sample_date_after).
+        :type filter_val: str, int
+        :param resource: MERMAID project entity resources.
+        :return: projects path, returns None if invalid project name or id.
+        :rtype: str, None
+        """
+        # Generate path.
+        path = None
+        if resource:
```

Comment:
See comment below. I think you can start with `path = ""` and then build each piece of the path, thus avoiding nested conditionals (and shortening code). Possibly:

```
path = ""
if id or name:
    path = "projects/"
p_id = id
if name:
    p_id = self.get_project_id(name=name)
if p_id:
    path += f"{p_id}/"
if resource:
    path += f"{resource}/"
```

But note a subtle difference: here, if `name` is specified, it trumps `id`. If your function accepts either (and I'm not sure it wouldn't be a better idea just to have two functions, instead), you'll always have the problem of what to do if both name and id are provided. I think the docstring help above should make clear which one trumps the other, rather than that they're both simply optional.